Integration tests by compiling and calling Ghostferry from Ruby

shuhaowu opened the pull request:
Instead of running integration tests directly in Go, this PR enables us to run integration tests by calling Ghostferry from Ruby as a subprocess. This is needed because it is _not possible_ for the the Go code to interrupt and resume itself.

Given that the Ruby integration test code is also much cleaner than the Go equivalent and suffer less issues such as goroutine leakage, we should prefer writing integration tests in Ruby from now on.

The Ghostferry subprocess right now is a separate binary from copydb/sharding. This decision was made because the tests needs to perform some actions (such as interrupting Ghostferry or injecting some data) in the middle of the run at strategic and reproducible times(1) and the base Ghostferry framework doesn't support it for now. In order to accomplish this, bidirectional communication needs to exists between the Ruby and Go code and this is done via a WEBrick::HTTPServer on the local machine. 

Using this server, Ghostferry can send a message and pause until a reply is sent from the server. The pause allows race injection and other test code to execute.

All of this is wrapped up in the class `GhostferryHelper::Ghostferry`. To use it in tests, one can extend from `GhostferryTestCase`, where other helpers like the database connection, a parallel data writer, also exists.

A test will look something like the following:

```
  def test_copy_data_with_writes_to_source
    seed_simple_database_with_single_table

    datawriter = new_source_datawriter
    ghostferry = new_ghostferry(MINIMAL_GHOSTFERRY)

    start_datawriter_with_ghostferry(datawriter, ghostferry)
    stop_datawriter_during_cutover(datawriter, ghostferry)

    ghostferry.run
    assert_test_table_is_identical
  end
```

shuhaowu commented:
It would be great if I can get some reviews on this PR. It's a fairly independent piece that we'll need regardless what approach we take to implement the rest of #75 as we'll need a mechanism to test.

At least half of this PR is reasonably trivial: direct translation from Go (DataWriter, TrivialIntegrationTest, TestCase), glue code to work with database (DbManager). The other half is a simple implementation of a Ghostferry with unix socket for status reporting.

*For future note: the integrationferry.Ferry.Main could be a method we'll eventually port back into ghostferry.Ferry, to simplify how Ghostferry application can be run as per an idea @pushrax had sometime ago.*

hkdsun commented on test/integration/go/integrationferry/integrationferry.go:
```diff
@@ -0,0 +1,238 @@
+	*ghostferry.Ferry
+}
+
+// =========================================
```
I also don't understand the need for this custom communication protocol.

Another option is to use the already existing `HTTPCallback` facility, similar to how the `sharding` package uses it in `sharding/testhelpers/unit_test_suite.go` and how the shop mover integration works (see `ghostferry.rb` in core).

hkdsun commented:
By the way, you may want to check out and possibly port some of the shop mover tests from core

shuhaowu commented on test/integration/go/integrationferry/integrationferry.go:
```diff
@@ -0,0 +1,238 @@
+	StatusDone                   string = "DONE"
+
+	// Could be sent by multiple goroutines in parallel
+	StatusBeforeRowCopy     string = "BEFORE_ROW_COPY"
```
They will be useful. For example, if we want to terminate Ghostferry in a test after copy has begun (in a later PR):

```
@ghostferry.on_status(Ghostferry::Status::AFTER_ROW_COPY) do
  batches_written += 1
  if batches_written >= 2
    @ghostferry.send_signal("TERM")
  end
end
```

Furthermore it is the intention to eventually port all existing Ghostferry integration tests into this framework. Lots of those Go tests uses the same statuses.

shuhaowu commented:
I've updated this PR to follow more of a Ruby convention:

- [x] Uses `bundle exec rake test`
- [x] `require "test_helper"`
- [x] Removed `GhostferryIntegration` module and moved functionality into `*Helper` modules that are included in `GhostferryTestCase`. 
- [x] Avoided implicit instance variables like `@ghostferry` and force tests to create their own instances of ghostferry via `ghostferry = new_ghostferry(minimal); ghostferry.run`.
- [x] Avoided ambiguously named functions like `use_datawriter` in favour of `start_datawriter_with_ghostferry(datawriter, ghostferry); stop_datawriter_during_cutover(datawriter, ghostferry)`
- [x] Used `WEBrick::HTTPServer` instead of UNIX socket
- [x] Moved all go test code from `test/*.go` into `test/go/*.go`

shuhaowu commented on test/helpers/data_writer_helper.rb:
```diff
@@ -0,0 +1,132 @@
+    end
+  end
+
+  class DataWriter
```
Much of this is translated directly from data_writer.go. The goal is eventually we can use Ruby tests only.

shuhaowu commented on test/helpers/db_helper.rb:
```diff
@@ -0,0 +1,150 @@
+    [source_metrics, target_metrics]
+  end
+
+  def table_metric(conn, table, sample_id: nil)
```
Not a big fan of this name, any ideas?

shuhaowu commented on test/test_helper.rb:
```diff
@@ -0,0 +1,109 @@
+    assert_equal(
+      source[DEFAULT_FULL_TABLE_NAME][:sample_row],
+      target[DEFAULT_FULL_TABLE_NAME][:sample_row],
+    )
```
The checksum is likely sufficient and the row count (especially since you can have a checksum of `0` if a table is empty), I'm just being defensive. I'll remove this line.
